### ISSUE TITLE
Make live changes from command liine (requires client-server architecture)

### ISSUE THREAD

**Issue by heyakyra** [Ongy/waymonad#17]
There are projects like [herbstluftwm](https://www.herbstluftwm.org/) and [bspwm](https://github.com/baskerville/bspwm) that use the client-server architecture so that live changes can be made from the terminal. This would be great for waymonad if it doesn't already do this!

**Comment by Ongy** [comment]
tbh. I'm not sure yet how much I want to do that.

The main reasons I think wayland will be better than X are better protection against snooping input and for waymonad, that we are "in control".
Depending on what's kept in a client, both of these properties have to be relaxed to a certain degree.

What I originally planed (though got more difficult by a decision along the way) was to allow a layouting client, to make those exchangable at runtime.

In general I don't think a client/server model makes much sense for this project, though there may possibilities for this that I don't see atm.

If you could be more specific, what should be configurable at runtime I can probably give a more adequat answer.
(It would help if you are familiar with xmonad, since I keep a certain amount of ideas from it)

**Comment by ghost** [comment]
The server/client architecture doesnt prohibit capabilities.
Of course, it's not a mainstream concept yet. Capnproto comes to mind, but the haskell implementation doesnt have rpc (yet). Id be intersted to work on this; I'm looking for a WM that would integrate with an RPC-based system (a-la distributed computer architecture -> throw the shell in the garbage and replace it with structured IPC). I haven't found anything yet, and used xmonad for years.

**Comment by ghost** [comment]
> In general I don't think a client/server model makes much sense for this project

That would mean it could never unify two hosts into a single workspace, which is where multi-monitor is going (each screen gets its own ressources, scales much better)

**Comment by Ongy** [comment]
> That would mean it could never unify two hosts into a single workspace, which is where multi-monitor is going (each screen gets its own ressources, scales much better)

Have you taken a look at recent graphics hardware? We are going towards 6+ Outputs on a single card.
At least on the consumer hardware.

**tl;dr:** I think you have a slightly wrong idea of what's possible on "plain" wayland.

What is supported (or should be, not 100% sure on the current situation and havne't tested in waymonad) is to use multiple cards in a single system.

What isn't supported by wayland at all: Clients that can't use shared mem with the compositor
This just isn't in the protocol. **Every** mechanism to send a rectangle of pixels from a client ot the server requires this (be it CPU-Ram or GPU [DMA-BUF]).
So your idea has to be supported as some kind of extension. (Maybe over some RDP-like connection?)

What I'll fully support and think should be doable with quite doable effort is a system ala. synergy that allows to use input devices connect to one system on all mostly seamless.
Moving an application between outputs that live on different hosts is quite a bit more ambitious.
If you really want to do this, I don't mind helping out building the roadmap of necessary tools/mechanisms.
And I'd gladly provide waymonad as playground for this, if it's properly layed out.

**Comment by ghost** [comment]
Thank you for explaining this.

> Moving an application between outputs that live on different hosts is quite a bit more ambitious.

That's the kind of stuff I plan to do, but certainly not with the current userland architecture (edit: obviously, the applications' state would be on all hosts then). Hence my comment here. A client/server separation would make it easier to integrate our ideas with a given WM. So I what I'm trying to say is that it would allow a lot of possibilities in the not-so-distant future.

On the powerful GPU, I'm not into high energy loss ideas. I see electricity prices going up 50-100x within a generation. Think people spending 15% of their salaries to power 3 or 4 very low-powered computers (a client, and a few thin servers). Those GPU's wouldnt be used in that scenario. But I could be wrong, and someone might control nuclear fusion.

Hopefully one day we can play with this. I'll give waymonad a spin soon. Thank you

**Comment by SebMdB** [comment]
> I see electricity prices going up 50-100x within a generation. Think people spending 15% of their salaries to power 3 or 4 very low-powered computers (a client, and a few thin servers).

That's unlikely. The world is switching to renewable energy sources that are getting cheaper than fossil fuels. 
The increase in electricity price is generally caused by two temporary factors (depending on your country): carbon pricing and investments in renewables.
However, the cost of production of those high end GPUs will probably increase for environmental reasons, which is in agreement with your reasoning.

**Comment by ghost** [comment]
I'v never heard of cheaper energy source than oil and gas. Former energy futures trader here. Cheap oil is an input into 100 percent of our advanced techs. Scarcer oil in the future means more expensive wind turbines. BTW, only plants are renewable; wind turbines and solar panels degrade with use, and nobody has found a way to grow magnets, copper or rare earth metals. Sorry if this leaves you in a state of lament for you future, but you sounded too intelligent for me to leave you in such a state of ignorance about energy. Our biggest problem is going to be replacing diesel pumps; motors take way too much electricity to be seriously considered by anyone who has any brain. So muscle energy is the only candidate left, along with mechanical wind/hydro which are very location-bound. Hence the current model I have for the price at which electricity settles; a man peddaling to produce power would need enough to sustain himself, and he can hope to produce at most a couple kWhr's in a day's work. Just because the chinese dictators are subsiding solar panels at a loss while completly ignoring the environmental costs doesnt mean they're cheaper than oil! Anyhow, I broke my rule again; Im not supposed to do this since nobody ever gets it. Or almost. See? I'm really an optimist!

**Comment by Profpatsch** [comment]
@lifranc I think your account has been compromised.

**Comment by ghost** [comment]
@Profpatsch I dont think so. But if you see me giving free advice again then this account (not my account; github owns github) is probably compromised. I just took an oath never to give my invaluable advice for free again. Thanks for your consideration!

**Comment by Profpatsch** [comment]
Re-reading the whole thread, this was actually on-topic. smh

I think you actually have a point about today’s “renewable energy” sources. But we had a pretty good solution 40 years ago: nuclear fission (with potential for fusion down the road). If only we had kept up research instead of panicking.

**Comment by ghost** [comment]
@Profpatsch The only possible outcome of nuclear energy is the end of all organic life.
